cmdfs.1
.TH cmdfs 1 "24 July 2012"
.SH NAME
cmdfs \(hy command file system
.SH SYNOPSIS
.B cmdfs
<\fIsource-dir\fR> <\fImount-dir\fR> [\fIoptions\fR]
.SH DESCRIPTION
\fIcmdfs\fP is a FUSE virtual filesystem that applies an arbitary
filter command to selected files in a source directory tree to create
the destination files. It includes configurable caching and monitoring
of the source tree to limit CPU use and prefetch data.
.SH OPTIONS
\fIcmdfs\fP accepts the following options:
.TP 8
.B  \-o command=<\fIshell command\fR>
The command to run to generate the file's content [default: cat]
.TP 8
.B  \-o extension=\fIext1\fR[;\fIext2\fR[;...]]
Specify matching file extension(s) to which command is applied
.TP 8
.B  \-o path-re=<\fIregular expression\fR>
Specify regexp of pathnames to which command is applied
.TP 8
.B  \-o exclude-re=<\fIregular expression\fR>
Specify regexp of pathnames to which command is never applied
.TP 8
.B  \-o mime-re=<\fIregular expression\fR>
Specify regexp of mime types (as returned by \fIfile
  -b --mime-type\fP) to which command is applied
.TP 8
.B  \-o link-thru
Pass unmatched files through to filesystem as symbolic links [default:
files not listed or accessible]
.TP 8
.B  \-o hide-empty-dirs
Don't show directories that have no matching descendant files
.TP 8
.B  \-o monitor
Watch source directory tree for new files and cache them as they
appear. New directories will be monitored [default: not monitored]
.TP 8
.B  \-o cache-dir=<\fIdirectory\fR>
Directory to save cache files [default:
/usr/local/var/cache/cmdfs/<\fIuser\fR>/<\fIsource-dir\fR>]
.TP 8
.B  \-o cache-size=<\fIsize in Mb\fR>
Size to attempt to limit cache directory to. Files removed on
least-recently access basis. [default: no limit]
.TP 8
.B  \-o cache-entries=<\fIcount\fR>
Number of entries to attempt to limit cache directory to. Files
removed on least-recent access basis. [default: no limit]
.TP 8
.B  \-o cache-expiry=<\fItime in seconds\fR>
Items in cache older than this will be replaced when read
[default: never expires]
.TP 8
.B  \-o stat-pass-thru
Don't force a stat() call to create cached file from source, just use
the source file as the target if not available.  Saves slow
directory listings if accurate sizes etc. not required.
[default: stat() calls create a cached file]
.SH EXAMPLES
Given a source tree, that includes, say, jpg images, we can generate a view
filesystem which contains the same files resized to email size.
The filter command is set up as a fileystem option at mount time. In this case
our filter option might be something like (using ImageMagick's convert
command):
.TP
	command=convert - -resize 20% -
.PP
Note we follow the UNIX convention of taking the standard input and filtering
to standard output. Being a FUSE filesystem, the child process will be run
under the group/user of the owner. The command is run within a shell (/bin/sh)
process, as if started by /bin/sh -c '<command>'
Obviously there'll be files in the source tree which we dont want to apply the
filter to. In this case we identify files of interest by extension:
.TP
	extension=jpg;gif;png
.PP
To mount the filesystem as above for a particular user:
.TP
	cmdfs /home/bob/images /mnt/myphotos "-ocommand=convert - -resize 20% -,extension=jpg;gif;png"

.SS Selection Filters

As well as just matching on extension, files of interest can be selected by
mime type (cmdfs uses the file command to determine the
mimetype) or matching regexps:
.TP
	mime-re=image/*
.PP
Or just a plain path regexp (this allows us to include whole directory subtrees
etc
.TP
	path-re=.*/my-images/
.PP
Or exclude patterns:
etc
.TP
  exclude-re=.*/working-files/;/secretstuff
.PP
These options can be combined to get the required filter selection.

.SS Unmatched Files
By default, files which don't match a filter won't appear in the filesystem
view. We may want to see the original files though, so the
.TP
	link-thru
.PP
option may be specified. This will generate symbolic links to the original
files.

.SS Monitoring
The image resize example is a good example of an application where we might
want the files in the source filesystem to be filtered as they are added,
rather than just on-demand. Cmdfs uses inotify to monitor all the directories
in the tree, and automatically monitors new subdirectories as they are added:
.TP
	monitor
.PP
When you copy the latest batch of files from your camera the resized versions
will automatically be generated.
.PP
Note that inotify has a limit to the number of directories that can be
monitored. On mounting, cmdfs will attempt to use all the handles it can get.
When it detects directories being removed, it will attempt to reclaim those
released handles for directories it was previously unable to cover.

.SS Caching
cmdfs only recreates view files if the modification date changes on the source.
All files created using the filter are held in a cache directory, with a name
created as a hash from the full path. By default this directory is located in /
tmp/cmdfs-cache.<user name>, but can be changed:
.TP
	cache-dir=<dir>
.PP
Note that some systems will clear the /tmp directory tree between reboots,
which may or may not be desired.
The size of the cache directory can be limited using:
.TP
	cache-size=<size in mb>
and/or
.TP
\&  cache-entries=<number of files>
.PP
The least recently accessed cached files will be removed in the background to
maintain these limits.
Expiry on cached files can be specified with:
.TP
	cache-expiry=<time in secs>
.PP
Cached files will be recreated if the are older than this.

.SS Mounting With fstab

To aid in mounting directories for multiple users, cmdfs accepts the source
base directory as the first argument so it can be mounted using the
conventional mount.fuse script from fstab. This allows supplying cmdfs#<dir> as
the fs_spec (see man 5 fstab). The user_allow_other option needs to be enabled
in /etc/fuse.conf to allow normal users to mount (see fuse.conf). An example
fstab entry:

.DS
  cmdfs#/media/myphotos /home/bob/images fuse user,allow_other,command=convert\\040-\\040-resize\\04020%\\040-,monitor,cache-size=500,extension=jpg;gif;png 0 0
.DE

Note that the spaces in the command need to be escaped using \\040 for /etc/
fstab to parse correctly

.SS Example 2: Dynamically Generated Files

The command used to generate the output does not need to be a simple stdin-
>stdout filter, it can generate its output from whatever source it chooses. Of
course this is of limited utility unless you can specify what command/
parameters to run (otherwise every file in the filesystem would have the same
content!).
The shell command may therefore contain a substitution token, %f, which will be
replaced with the full path/filename of the input file. If, for example, the
command is simply '%f', cmdfs will attempt to execute the file (it should have
+x permission to allow this)
This effectively gives a 'CGI' type function to the filesytem. e.g. If may
contain a script:
.DS
  % cat > /home/bob/showps
  #!/bin/sh
  ps -ef
  ^D
  % chmod +x /home/bob/showps
  % cmdfs /home/bob /tmp/test "-opath-re=.*,command=%f,cache-expiry=5"
  % cat /tmp/test/showps
  UID        PID  PPID  C STIME TTY          TIME CMD
  root         1     0  0 Feb13 ?        00:00:01 /sbin/init
  root         2     0  0 Feb13 ?        00:00:00 [kthreadd]
  root         3     2  0 Feb13 ?        00:00:00 [migration/0]
  root         4     2  0 Feb13 ?        00:00:09 [ksoftirqd/0]
  root         5     2  0 Feb13 ?        00:00:00 [watchdog/0]
  \...
.DE
.PP
Note that the cache expiry time in the example has been set at 5 secs - this
means the ps results will be re-created if older than 5 sesonds.


.IP \(bu 4
If the executable returns a non-zero exit status, cmdfs will assume that it
  has failed (will appear as 'No such file or directory').
.IP \(bu 4
It is not recommend to use the %f facility to modify the source file - this
  will invalidate the cache immediately resulting in a recreation of the output
  on next read (unless this is intended)
.IP \(bu 4
cmdfs also sets an environment variable INPUT_FILE in the child command
  process, which contains the source file name, which may be useful if the
  eventual executable is being called via a parent script.


.Ve
.SH "SEE ALSO"
.BR fusermount (1),
.SH AUTHOR
Mike Swain <mike@hiko.co.nz>


ChangeLog
27-01-2010 Mike Swain <mike@yagz.org.uk>
 * Initial release 0.1
 
30-01-2010 Mike Swain <mike@yagz.org.uk>
 * Release 0.2
	 Remove base-dir option - specify with first argument for fstab
	 Code tidy
	 Bug, inotify event name not null terminated
	 Update help and README


02-14-2010 Mike Swain <mike@yagz.org.uk>
 * Release 0.3
	 Add cache_expiry option
	 Add %f dynamic file generation
	 Add INPUT_FILE env var
	 Update help and README
 
03-24-2010  Mike Swain <mike@yagz.org.uk>
 * Release 0.4
	 Add hide-empty-dirs option
	 Add test suite
	 Various bugs fixed
	 Caching changed to not use md5 hashing (for performance). Names are as is with dirs flattened.
	 Default Cache location changed to [/usr/local]/var/log
	 Update help and README

16-07-2012  Mike Swain <mike@hiko.co.nz>
 * Release 0.5
	 Add stat-pass-thru
	 Add test to test suite and improve tests
	 Fix Amd64 build
	 Various bugs fixed
	 Add more logging (e.g. dumps options at start of session)
	 

src/cmdfs.c
/**
* Required by fuse.h
*/
#define FUSE_USE_VERSION 26

#include "config.h"
#include "cmdfs.h"

#include <errno.h>
#include <fcntl.h>
#include <fuse.h>
#include <fuse/fuse_opt.h>
#include <dirent.h>
#include <stdint.h>
#include <pthread.h>


options_t options = {
	.hide_empty_dirs = 0,
	.link_thru = 0,
	.stat_pass_thru = 0,
	.monitor = 0,
	.mount_dir = NULL,
	.base_dir = NULL,
	.cache_dir = NULL,
	.cache_entries = 0,
	.cache_size = 0,
	.cache_expiry = -1L,
	.cache_max_wait = 600,
	.command = NULL,
	.fnmatch = NULL,
	.fnmatch_c = 0,
	.path_regexps = NULL,
	.path_regexp_cnt = 0,
	.exclude_regexps = NULL,
	.exclude_regexp_cnt = 0,
	.mime_regexps = NULL,
	.mime_regexp_cnt = 0
};

monitor_t *monitor = NULL;
cleaner_t *cleaner = NULL;


static int is_empty_visitor( const dir_info *visit, void *data ) {
	int rv = 0;
	if ( S_ISREG(visit->mode) ) {
		if ( options.link_thru ) {
			rv = 1; // there is a regular file and linked thru, so not empty
		}
		else {
				vfile_t *f = file_create_from_src(visit->path);
				rv = file_get_command(f) != NULL; // virtual file is to be generated so not empty
				file_destroy(f);
		}
	}
	return rv;
}
static int is_empty(const char *dirpath) {
	return dir_visit(dirpath,-1,is_empty_visitor,NULL) == 0;
}


int cmdfs_open(const char *path, struct fuse_file_info *info) {
	info->fh = (uint64_t)(long)file_create_from_dst(path);

	return 0;
}

int cmdfs_getattr(const char *path, struct stat *st) {
	vfile_t *f = file_create_from_dst(path);
	int rv = 0;
	const char *src = file_get_src(f);
	if ( stat(src,st))
		rv = -errno;
	else if (S_ISREG(st->st_mode)) {
		if ( file_get_command(f)) {
			// is covered by command
		  const char *cacheName = file_get_cached_path(f);
		  struct stat dststat;
		  int cacheExists = cacheName && !stat(cacheName,&dststat);
		  int cacheIsFile = cacheExists && S_ISREG(dststat.st_mode);

		  if (options.stat_pass_thru && !cacheIsFile) { // either can pass stat through and uncached, or stat of cached file failed
		    if ( stat(src, st))
		      rv = -errno;
		  } else {
		    if ( !cacheExists && stat(file_encache(f),&dststat)) // okay, wasn't cached before so cache and stat
		      rv = -errno;
		    else {
		      st->st_size = dststat.st_size;
		      st->st_mode &= S_IFREG | 0444; // always readonly
		    }
		  }
		}
		else if (options.link_thru) {
			// Return read only link
			st->st_mode = S_IFLNK | (st->st_mode & 0444);
			st->st_nlink = 2;
			st->st_size = strlen(path);
		}
		else
			// Pretend nothing's there
			rv = -ENOENT;
	}
	else if (S_ISDIR(st->st_mode) && options.hide_empty_dirs && is_empty(src)) {
		// Pretend nothing's there - directory and subdirs are empty of matching files
		rv = -ENOENT;
	}
	file_destroy(f);
	return rv;
}

struct readdir_collector {
	fuse_fill_dir_t fill;
	off_t offset;
	struct fuse_file_info *info;
	int pos;
	void *buf;
};
static int readdir_visitor( const dir_info *visit, void *data ) {
	//log_debug(visit->path);
	struct readdir_collector *c = (struct readdir_collector *)data;
	if (++c->pos > c->offset ) {
		int isParent = !strcmp(visit->name,"/") || !strcmp(visit->name,".."); // special case '..' looks in mount's parent
    const char *cpath = visit->path;
		char parent[strlen(options.mount_dir)+4];
		if ( isParent ) {
			sprintf(parent,"%s/..",options.mount_dir);
			cpath = parent;
		}
		if (isParent || (S_ISDIR(visit->mode) && // its a dir
				!(options.hide_empty_dirs && is_empty(cpath))) || // its empty and we're hiding empties
				options.link_thru) { // or linking thru
			if (c->fill(c->buf,visit->name,NULL,c->pos))
				return -1; // break out
		}
		else if (!options.link_thru && S_ISREG(visit->mode)) {
			// we don want to show if it doesn't match command filter
			vfile_t *f = file_create_from_src(cpath);
			if ( file_get_command(f) && c->fill(c->buf,visit->name,NULL,c->pos)) {
					file_destroy(f);
					return -1;
			}
			file_destroy(f);
		}
	}
	return 0;
}

int cmdfs_readdir(const char *path, void *buf, fuse_fill_dir_t fill, off_t offset, struct fuse_file_info *info) {
	vfile_t *d = file_create_from_dst(path);
	int rv = 0;
	const char *src = file_get_src(d);
	if (options.hide_empty_dirs && is_empty(src)) {
		rv = -ENOENT; // dir hidden
	}
	else  {
		struct readdir_collector collect = {
			.fill = fill,
			.offset = offset,
			.info = info,
			.buf = buf
		};
		if ( dir_visit(src,0,readdir_visitor,&collect) < 0 )
			rv = -errno;
	}
	file_destroy(d);
	return rv;

}

int cmdfs_release(const char *path, struct fuse_file_info *info) {
	vfile_t *f = (vfile_t *)(long)info->fh;
	if ( f ) {
		file_destroy(f);
		info->fh = 0;
	}
	return 0;

}

int cmdfs_readlink(const char *path, char *buf, size_t size) {
	vfile_t *l = file_create_from_dst(path);
	strcpy(buf,file_get_src(l));
	file_destroy(l);
	return 0;
}
int cmdfs_read(const char *path, char *buf, size_t size, off_t offset, struct fuse_file_info *info) {
	vfile_t *f = (vfile_t *)(long)info->fh;
	if ( f ) {
		lseek(file_get_handle(f),offset,SEEK_SET);
		return read(file_get_handle(f), buf, size);
	}
	else
		return -EIO;
}

void *cmdfs_init(struct fuse_conn_info *conn) {
	if ( options.monitor ) {
		monitor = monitor_create(options.base_dir,options.mount_dir);
		log_debug("monitor thread created");
	}
	if ( options.cache_entries || options.cache_size || options.cache_expiry > 0 ) {
		cleaner = cleaner_create(options.cache_dir,options.cache_size,options.cache_entries, options.cache_expiry);
		log_debug("cleaner thread created");
	}
	return NULL;
}

void cmdfs_destroy(void *p) {
	if ( monitor ) {
		monitor_destroy(monitor);
		log_debug("monitor thread destroyed");
	}
	if ( cleaner ) {
		cleaner_destroy(cleaner);
		log_debug("cleaner thread destroyed");
	}
	log_debug("end of session");
}
static struct fuse_operations cmdfs_operations = {

	.init = cmdfs_init,
	.getattr   = cmdfs_getattr,
    .readdir   = cmdfs_readdir,
    .open   = cmdfs_open,
    .read   = cmdfs_read,
    .release = cmdfs_release,
    .readlink = cmdfs_readlink,
    .destroy = cmdfs_destroy
};


/** macro to define options */
#define CMDFS_OPT_KEY(t, p, v) { t, offsetof(options_t, p), v }

/** keys for FUSE_OPT_ options */
enum
{
   KEY_HELP,
   KEY_VERSION,
   KEY_EXTENSION,
   KEY_PATH_RE,
	 KEY_EXCLUDE_RE,
   KEY_MIME_RE
};

struct fuse_opt cmdfs_opts[] = {

	CMDFS_OPT_KEY("link-thru", link_thru, 1),
	CMDFS_OPT_KEY("nolink-thru", link_thru, 0),
	CMDFS_OPT_KEY("stat-pass-thru", stat_pass_thru, 1),
	CMDFS_OPT_KEY("nostat-pass-thru", stat_pass_thru, 0),
	CMDFS_OPT_KEY("hide-empty-dirs", hide_empty_dirs, 1),
	CMDFS_OPT_KEY("nohide-empty-dirs", hide_empty_dirs, 0),
	CMDFS_OPT_KEY("monitor",   monitor, 1),
	CMDFS_OPT_KEY("nomonitor",   monitor, 0),

	CMDFS_OPT_KEY("cache-dir=%s",   cache_dir, 0),
	CMDFS_OPT_KEY("cache-size=%lu",   cache_size, 0),
	CMDFS_OPT_KEY("cache-entries=%lu",   cache_entries, 0),
	CMDFS_OPT_KEY("cache-expiry=%lu",   cache_expiry, 0),
	CMDFS_OPT_KEY("command=%s",   command, 0),
	FUSE_OPT_KEY("extension=%s",KEY_EXTENSION),
	FUSE_OPT_KEY("path-re=%s",KEY_PATH_RE),
	FUSE_OPT_KEY("exclude-re=%s",KEY_EXCLUDE_RE),
	FUSE_OPT_KEY("mime-re=%s",KEY_MIME_RE),

	FUSE_OPT_KEY("-V",             KEY_VERSION),
	FUSE_OPT_KEY("--version",      KEY_VERSION),
	FUSE_OPT_KEY("-h",             KEY_HELP),
	FUSE_OPT_KEY("--help",         KEY_HELP),

	FUSE_OPT_END
};

static char *get_regerror (int errcode, regex_t *compiled) {
	size_t length = regerror (errcode, compiled, NULL, 0);
	char *buffer = (char *)malloc (length);
	(void) regerror (errcode, compiled, buffer, length);
	return buffer;
}

static int cmdfs_opt_proc(void *data, const char *arg, int key, struct fuse_args *outargs)
{
	 char *val = strchr(arg,'=');
     switch (key) {
     case KEY_HELP:
             fprintf(stderr,
                     "usage: %s source-dir mountpoint [options]\n"
                     "\n"
                     "general options:\n"
                     "    -o opt,[opt...]  mount options\n"
                     "    -h   --help      print help\n"
                     "    -V   --version   print version\n"
                     "\n"
                     "Cmdfs options:\n"
                     "    -o command=<shell command> (dd)\n"
                     "    -o extension=ext1[;ext2[;...]]\n"
                     "    -o path-re=<regular expression>\n"
										 "    -o exclude-re=<regular expression>\n"
                     "    -o mime-re=<regular expression>\n"
            		 "    -o [no]link-thru (nolink-thru)\n"
                         "    -o [no]stat_pass_thru (nostat_pass_thru)\n"
            		 "    -o [no]hide-empty-dirs (nohide-empty-dirs)\n"
            		 "    -o [no]monitor (nomonitor)\n"
            		 "    -o [no]stat-pass-thru (stat-pass-thru)\n"
            		 "    -o cache-dir=<dir> (%s/<user>/<source-dir>)\n"
            		 "    -o cache-size=<size in Mb> (no limit)\n"
            		 "    -o cache-entries=<count> (no limit)\n"
            		 "    -o cache-expiry=<time in secs> (no expiry)\n"
                     , outargs->argv[0], CACHE_ROOT);
             fuse_opt_add_arg(outargs, "-ho");
             fuse_main(outargs->argc, outargs->argv, &cmdfs_operations, NULL);
             exit(1);

     case KEY_VERSION:
             fprintf(stderr, "%s\n", PACKAGE_STRING);
             fuse_opt_add_arg(outargs, "--version");
             fuse_main(outargs->argc, outargs->argv, &cmdfs_operations, NULL);
             exit(0);
     case KEY_EXTENSION:
    	if (val && strlen(val)>0) {
    		char *exts = strdup(val+1);
    		char *save_ptr = NULL, *ext;
    		for ( ext = strtok_r(exts,";",&save_ptr); ext != NULL; ext = strtok_r(NULL,";",&save_ptr)) {
    			char *expr;
    			int err;
    			options.path_regexps = realloc(options.path_regexps,sizeof(regex_t)*++options.path_regexp_cnt);
    			if ( asprintf(&expr,".*/.*\\.%s",ext) < 0 )
    				exit(0);
    			if ((err=regcomp(options.path_regexps+(options.path_regexp_cnt-1), expr,REG_ICASE|REG_NOSUB))) {
    				log_error("Error compiling regex: %s",get_regerror(err,options.path_regexps+(options.path_regexp_cnt-1)));
    				return 1;
    			}
    			free(expr);
    		}
    		free(exts);
    		log_debug("extension: %s\n",val+1);
			return 0;
    	}
     case KEY_PATH_RE:
    	if (val && strlen(val)>0) {
			int err;
			options.path_regexps = realloc(options.path_regexps,sizeof(regex_t)*++options.path_regexp_cnt);
			if ((err=regcomp(options.path_regexps+(options.path_regexp_cnt-1), val+1,0))) {
				log_error("Error compiling path regex: %s",get_regerror(err,options.path_regexps+(options.path_regexp_cnt-1)));
				return 1;
			}
			log_debug("path-re: %s",val+1);
			return 0;
    	}
			case KEY_EXCLUDE_RE:
     	if (val && strlen(val)>0) {
 			int err;
 			options.exclude_regexps = realloc(options.exclude_regexps,sizeof(regex_t)*++options.exclude_regexp_cnt);
 			if ((err=regcomp(options.exclude_regexps+(options.exclude_regexp_cnt-1), val+1,0))) {
 				log_error("Error compiling exclude regex: %s",get_regerror(err,options.exclude_regexps+(options.exclude_regexp_cnt-1)));
 				return 1;
 			}
 			log_debug("exclude-re: %s",val+1);
 			return 0;
     	}
     case KEY_MIME_RE:
    	if (val && strlen(val)>0) {
			int err;
			options.mime_regexps = realloc(options.mime_regexps,sizeof(regex_t)*++options.mime_regexp_cnt);
			if ((err=regcomp(options.mime_regexps+(options.mime_regexp_cnt-1), val+1,0))) {
				log_error("Error compiling regex: %s",get_regerror(err,options.mime_regexps+(options.mime_regexp_cnt-1)));
				return 1;
			}
			log_debug("mime-re: %s",val+1);
			return 0;
    	}
	 case FUSE_OPT_KEY_NONOPT:
		 // base dir can be supplied as first argument (allows fstab config)
		 if (!options.base_dir) {
			 options.base_dir = canonicalize_file_name(arg);
			 return 0;
		 }
		 else if ( !options.mount_dir ) {
			 options.mount_dir = canonicalize_file_name(arg);
			 // drop thru as fuse picks this up
		 }
     }

     return 1;
}

void dump_options() {
	log_debug("base_dir: %s",options.base_dir);
	log_debug("mount_dir: %s",options.mount_dir);
	log_debug("cache_dir: %s",options.cache_dir);
	log_debug("cache_size: %lu",options.cache_size);
	log_debug("cache_expiry: %ld",options.cache_expiry);
	log_debug("monitor: %d",options.monitor);
	log_debug("link_thru: %d",options.link_thru);
	log_debug("hide_empty_dirs: %d ",options.hide_empty_dirs);
	log_debug("stat_pass_thru: %d",options.stat_pass_thru);
	log_debug("command: %s\n",options.command);
	for ( int i = 0; i < options.fnmatch_c; i++)
		log_debug("extension: %s\n",options.fnmatch[i]);

}

int main(int argc, char **argv) {

	log_debug("starting session");
	int ret =0;
	struct fuse_args args = FUSE_ARGS_INIT(argc, argv);

	if (fuse_opt_parse(&args, &options, cmdfs_opts, cmdfs_opt_proc) == -1) {
		/** error parsing options */
		goto exit;
	}

	// default, token substitute and canonicalize cachedir
	if ( !options.cache_dir && asprintf((char **)&options.cache_dir,"%s/%%u/%%b",CACHE_ROOT)<0) {
		log_error("allocate cache directory string");
		goto exit;
	}
	const char *tokens[] = {"%u","%b","%m",NULL };
	const char *values[] = { getlogin(),options.base_dir,options.mount_dir,NULL };
	const char *toksub = tokens_substitute(options.cache_dir,tokens,values);
	free((void*)options.cache_dir);
	if (!(options.cache_dir = makepath(toksub)) ) {
		log_error("Could not create/find cache directory %s (%s)", toksub, strerror(errno));
		goto exit;
	}
	free((void *)toksub);

	if ( options.command == NULL ) {
		options.command = strdup("dd"); // default just copy original file
	}

	dump_options();

	ret = fuse_main(args.argc, args.argv, &cmdfs_operations, NULL);
exit:
	fuse_opt_free_args(&args);
	return ret;
}
